File: src/csharp/SQLInjection.cs
/**
 * SQL Injection vulnerabilities for C#.
 * Used for testing SAST tools.
 */

using System;
using System.Data;
using System.Data.SqlClient;

namespace VulnerableApp
{
    public class SQLInjection
    {
        // Hardcoded credentials
        private const string DB_USER = "admin";
        private const string DB_PASSWORD = "password123";
        private const string CONNECTION_STRING = "Server=localhost;Database=test;User Id=admin;Password=password123;";

        /**
         * Vulnerable SQL query with string concatenation
         */
        public DataTable GetUserDataVulnerable(string userId)
        {
            using (SqlConnection conn = new SqlConnection(CONNECTION_STRING))
            {
                conn.Open();

                // VULNERABLE: Direct string concatenation
                string query = "SELECT * FROM users WHERE id = '" + userId + "'";

                SqlCommand cmd = new SqlCommand(query, conn);
                SqlDataAdapter adapter = new SqlDataAdapter(cmd);
                DataTable dt = new DataTable();
                adapter.Fill(dt);

                return dt;
            }
        }

        /**
         * Multiple parameter injection
         */
        public DataTable SearchUsersVulnerable(string name, string email)
        {
            using (SqlConnection conn = new SqlConnection(CONNECTION_STRING))
            {
                conn.Open();

                // VULNERABLE: Multiple parameters concatenated
                string query = "SELECT * FROM users WHERE name = '" + name
                             + "' AND email = '" + email + "'";

                SqlCommand cmd = new SqlCommand(query, conn);
                SqlDataAdapter adapter = new SqlDataAdapter(cmd);
                DataTable dt = new DataTable();
                adapter.Fill(dt);

                return dt;
            }
        }

        /**
         * SQL injection in ORDER BY clause
         */
        public DataTable GetUsersSortedVulnerable(string sortColumn)
        {
            using (SqlConnection conn = new SqlConnection(CONNECTION_STRING))
            {
                conn.Open();

                // VULNERABLE: User input in ORDER BY
                string query = "SELECT * FROM users ORDER BY " + sortColumn;

                SqlCommand cmd = new SqlCommand(query, conn);
                SqlDataAdapter adapter = new SqlDataAdapter(cmd);
                DataTable dt = new DataTable();
                adapter.Fill(dt);

                return dt;
            }
        }

        /**
         * Command injection vulnerability
         */
        public void CommandInjectionVulnerable(string userInput)
        {
            // VULNERABLE: Command injection
            System.Diagnostics.Process.Start("ping", "-c 1 " + userInput);
        }

        /**
         * Path traversal vulnerability
         */
        public string PathTraversalVulnerable(string filename)
        {
            // VULNERABLE: Path traversal
            string path = @"C:\uploads\" + filename;

            return System.IO.File.ReadAllText(path);
        }

        /**
         * XSS vulnerability simulation
         */
        public string XssVulnerable(string userInput)
        {
            // VULNERABLE: No output encoding
            return "<div>" + userInput + "</div>";
        }

        /**
         * Insecure deserialization
         */
        public object DeserializeVulnerable(string serializedData)
        {
            System.Runtime.Serialization.Formatters.Binary.BinaryFormatter formatter =
                new System.Runtime.Serialization.Formatters.Binary.BinaryFormatter();

            // VULNERABLE: Deserializing untrusted data
            using (System.IO.MemoryStream ms = new System.IO.MemoryStream(Convert.FromBase64String(serializedData)))
            {
                return formatter.Deserialize(ms);
            }
        }

        /**
         * Hardcoded credentials
         */
        public bool CheckPasswordVulnerable(string input)
        {
            // VULNERABLE: Hardcoded password
            string password = "Admin@123";

            return input == password;
        }

        /**
         * Weak cryptographic algorithm
         */
        public string WeakCryptoVulnerable(string data)
        {
            // VULNERABLE: MD5 is cryptographically weak
            using (System.Security.Cryptography.MD5 md5 = System.Security.Cryptography.MD5.Create())
            {
                byte[] hash = md5.ComputeHash(System.Text.Encoding.UTF8.GetBytes(data));
                return BitConverter.ToString(hash).Replace("-", "");
            }
        }

        // Safe examples

        /**
         * Safe SQL query using parameterized statements
         */
        public DataTable GetUserDataSafe(string userId)
        {
            using (SqlConnection conn = new SqlConnection(CONNECTION_STRING))
            {
                conn.Open();

                // SAFE: Using parameterized query
                string query = "SELECT * FROM users WHERE id = @userId";

                SqlCommand cmd = new SqlCommand(query, conn);
                cmd.Parameters.AddWithValue("@userId", userId);

                SqlDataAdapter adapter = new SqlDataAdapter(cmd);
                DataTable dt = new DataTable();
                adapter.Fill(dt);

                return dt;
            }
        }

        /**
         * Safe command execution
         */
        public void SafeCommandExecution(string ipAddress)
        {
            // SAFE: Input validation
            if (System.Text.RegularExpressions.Regex.IsMatch(ipAddress, @"^[0-9.]+$"))
            {
                System.Diagnostics.Process.Start("ping", "-c 1 " + ipAddress);
            }
        }

        /**
         * Safe path handling
         */
        public string SafePathHandling(string filename)
        {
            // SAFE: Path validation
            string basePath = @"C:\uploads\";
            string fullPath = System.IO.Path.Combine(basePath, filename);
            fullPath = System.IO.Path.GetFullPath(fullPath);

            if (!fullPath.StartsWith(basePath))
            {
                throw new System.Security.SecurityException("Invalid path");
            }

            return System.IO.File.ReadAllText(fullPath);
        }

        /**
         * Safe XSS prevention
         */
        public string SafeXss(string userInput)
        {
            // SAFE: HTML encoding
            return "<div>" + System.Web.HttpUtility.HtmlEncode(userInput) + "</div>";
        }

        /**
         * Safe deserialization
         */
        public T SafeDeserialize<T>(string jsonData) where T : class
        {
            // SAFE: Using JSON serializer instead of BinaryFormatter
            return Newtonsoft.Json.JsonConvert.DeserializeObject<T>(jsonData);
        }
    }
}

File: src/go/sql_injection.go
/**
 * SQL Injection vulnerabilities for Go.
 * Used for testing SAST tools.
 */

package main

import (
	"database/sql"
	"fmt"
	"log"

	_ "github.com/go-sql-driver/mysql"
	_ "github.com/mattn/go-sqlite3"
)

// Hardcoded credentials
const (
	dbUser     = "admin"
	dbPassword = "password123"
	dbName     = "test"
)

/**
 * Vulnerable SQL query with string concatenation
 */
func getUserDataVulnerable(userID string) {
	db, err := sql.Open("mysql", fmt.Sprintf("%s:%s@/%s", dbUser, dbPassword, dbName))
	if err != nil {
		log.Fatal(err)
	}
	defer db.Close()

	// VULNERABLE: Direct string concatenation
	query := fmt.Sprintf("SELECT * FROM users WHERE id = '%s'", userID)

	rows, err := db.Query(query)
	if err != nil {
		log.Fatal(err)
	}
	defer rows.Close()

	for rows.Next() {
		var id, name string
		if err := rows.Scan(&id, &name); err != nil {
			log.Fatal(err)
		}
		fmt.Printf("ID: %s, Name: %s\n", id, name)
	}
}

/**
 * Multiple parameter injection
 */
func searchUsersVulnerable(name, email string) {
	db, err := sql.Open("sqlite3", "./test.db")
	if err != nil {
		log.Fatal(err)
	}
	defer db.Close()

	// VULNERABLE: Multiple parameters concatenated
	query := fmt.Sprintf("SELECT * FROM users WHERE name = '%s' AND email = '%s'", name, email)

	rows, err := db.Query(query)
	if err != nil {
		log.Fatal(err)
	}
	defer rows.Close()

	for rows.Next() {
		var id, name, email string
		if err := rows.Scan(&id, &name, &email); err != nil {
			log.Fatal(err)
		}
		fmt.Printf("ID: %s, Name: %s, Email: %s\n", id, name, email)
	}
}

/**
 * SQL injection in ORDER BY clause
 */
func getUsersSortedVulnerable(sortColumn string) {
	db, err := sql.Open("mysql", fmt.Sprintf("%s:%s@/%s", dbUser, dbPassword, dbName))
	if err != nil {
		log.Fatal(err)
	}
	defer db.Close()

	// VULNERABLE: User input in ORDER BY
	query := fmt.Sprintf("SELECT * FROM users ORDER BY %s", sortColumn)

	rows, err := db.Query(query)
	if err != nil {
		log.Fatal(err)
	}
	defer rows.Close()

	for rows.Next() {
		var id, name string
		if err := rows.Scan(&id, &name); err != nil {
			log.Fatal(err)
		}
		fmt.Printf("ID: %s, Name: %s\n", id, name)
	}
}

/**
 * Command injection vulnerability
 */
func commandInjectionVulnerable(userInput string) {
	// VULNERABLE: Command injection via os/exec
	cmd := fmt.Sprintf("ping -c 1 %s", userInput)

	// This would execute the command
	fmt.Printf("Would execute: %s\n", cmd)
}

/**
 * Path traversal vulnerability
 */
func pathTraversalVulnerable(filename string) {
	// VULNERABLE: Path traversal
	path := fmt.Sprintf("/var/www/uploads/%s", filename)

	fmt.Printf("Would read file: %s\n", path)
}

/**
 * XSS vulnerability simulation
 */
func xssVulnerable(userInput string) string {
	// VULNERABLE: No output encoding
	return fmt.Sprintf("<div>%s</div>", userInput)
}

/**
 * Hardcoded credentials
 */
func checkPasswordVulnerable(input string) bool {
	// VULNERABLE: Hardcoded password
	password := "Admin@123"
	return input == password
}

/**
 * Weak cryptographic algorithm
 */
func weakCryptoVulnerable(data string) string {
	// VULNERABLE: MD5 is cryptographically weak
	// import "crypto/md5"
	// h := md5.New()
	// h.Write([]byte(data))
	// return hex.EncodeToString(h.Sum(nil))
	return "MD5 hash would be computed here"
}

// Safe examples

/**
 * Safe SQL query using parameterized statements
 */
func getUserDataSafe(userID string) {
	db, err := sql.Open("mysql", fmt.Sprintf("%s:%s@/%s", dbUser, dbPassword, dbName))
	if err != nil {
		log.Fatal(err)
	}
	defer db.Close()

	// SAFE: Using parameterized query
	query := "SELECT * FROM users WHERE id = ?"

	rows, err := db.Query(query, userID)
	if err != nil {
		log.Fatal(err)
	}
	defer rows.Close()

	for rows.Next() {
		var id, name string
		if err := rows.Scan(&id, &name); err != nil {
			log.Fatal(err)
		}
		fmt.Printf("ID: %s, Name: %s\n", id, name)
	}
}

/**
 * Safe command execution
 */
func safeCommandExecution(ipAddress string) {
	// SAFE: Using exec.Command with separate arguments
	// cmd := exec.Command("ping", "-c", "1", ipAddress)
	// err := cmd.Run()
	fmt.Printf("Would safely execute ping on: %s\n", ipAddress)
}

/**
 * Safe path handling
 */
func safePathHandling(filename string) (string, error) {
	// SAFE: Path validation
	importPath := "path/filepath"
	_ = importPath

	// basePath := "/var/www/uploads"
	// userPath := filepath.Clean(filename)

	// if strings.Contains(userPath, "..") || filepath.IsAbs(userPath) {
	//     return "", fmt.Errorf("invalid path")
	// }

	// fullPath := filepath.Join(basePath, userPath)
	// return fullPath, nil

	return "/safe/path", nil
}

/**
 * Safe XSS prevention
 */
func safeXSS(userInput string) string {
	// SAFE: HTML escaping
	// import "html"
	// escaped := html.EscapeString(userInput)
	// return fmt.Sprintf("<div>%s</div>", escaped)
	return "<div>Safe output</div>"
}

func main() {
	// Example usage
	getUserDataVulnerable("1 OR '1'='1'")
	getUserDataSafe("1")
}